projects/telemetry/src/nmea.c

```diff
@@ -0,0 +1,182 @@
+StatusCode nmea_valid(const uint8_t *to_check, size_t len) {
+  if (to_check == NULL) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Cannot pass NULL pointer as parameter\n");
+  }
+  if (len < 3) {
+    return status_msg(STATUS_CODE_UNKNOWN,
+                      "The length is too short to contain useful information!\n");
+  }
+  if (to_check[0] != '$') {
+    return status_msg(STATUS_CODE_UNKNOWN, "First character of a NMEA sentence should be a $\n");
+  }
+  if (to_check[1] != 'G' || to_check[2] != 'P') {
+    return status_msg(STATUS_CODE_UNKNOWN, "NMEA sentence should begin with GP\n");
+  }
+  if (!nmea_checksum_validate((char *)to_check, len)) {
+    return status_msg(STATUS_CODE_UNKNOWN, "Invalid checksum for NMEA message\n");
+  }
+  return STATUS_CODE_OK;
+}
+
+// Returns the NMEA sentence type
+StatusCode nmea_sentence_type(const uint8_t *rx_arr, size_t len, NMEA_MESSAGE_ID *result) {
+  if (result == NULL || rx_arr == NULL) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Cannot pass NULL pointers to this function\n");
+  }
+
+  // Array index 3 should be 0, so that it is a null terminated string
+  uint8_t message_id[4] = { 0 };
+  for (uint16_t i = 3; i < len && i < 6; i++) {
+    if (rx_arr[i] == ',') {
+      break;
+    }
+    message_id[i - 3] = rx_arr[i];
+  }
+
+  // Making sure array index 3 is \n
+  message_id[3] = '\0';
+
+  // Parsing which type of NMEA message this is
+  if (strcmp((char *)message_id, "GGA") == 0) {
+    *result = NMEA_GGA;
+    return STATUS_CODE_OK;
+  } else if (strcmp((char *)message_id, "GLL") == 0) {
+    *result = NMEA_GLL;
+    return STATUS_CODE_OK;
+  } else if (strcmp((char *)message_id, "GSA") == 0) {
+    *result = NMEA_GSA;
+    return STATUS_CODE_OK;
+  } else if (strcmp((char *)message_id, "GSV") == 0) {
+    *result = NMEA_GSV;
+    return STATUS_CODE_OK;
+  } else if (strcmp((char *)message_id, "RMC") == 0) {
+    *result = NMEA_RMC;
+    return STATUS_CODE_OK;
+  } else if (strcmp((char *)message_id, "VTG") == 0) {
+    *result = NMEA_VTG;
+    return STATUS_CODE_OK;
+  } else {
+    LOG_DEBUG("Unknown message type: %c%c%c", message_id[0], message_id[1], message_id[2]);
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Unknown NMEA message type\n");
+  }
+}
+
+StatusCode nmea_get_gga_sentence(const uint8_t *rx_arr, size_t len, nmea_gga_sentence *result) {
+  if (result == NULL || rx_arr == NULL) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Cannot pass NULL pointers to this function\n");
+  }
+  status_ok_or_return(nmea_valid(rx_arr, len));
+
+  // m_id will keep track of which sentence type we are currently operating on
+  NMEA_MESSAGE_ID m_id = 0;
+
+  status_ok_or_return(nmea_sentence_type(rx_arr, len, &m_id));
+
+  result->message_id = m_id;
+
+  // Parse message_id below
+
+  char temp_buf[s_nmea_message_num_fields[NMEA_GGA]][10];
+
+  prv_split_nmea_into_array(rx_arr, len, s_nmea_message_num_fields[NMEA_GGA], 10, temp_buf);
```

Comment:
If your data fields are always split by commas, why not use strtok or your own version? (replace commas with null terminators)

Reply:
Will likely write my own function because `strtok` requires the string to be null terminated, and the parser requires the array to not be null terminated (since the GPS sends the data this way)

Reply:
The UART handler automatically terminates the array. See line 118/119 in `uart.c`.

Reply:
Thanks, I'll do it this way instead

Reply:
So it turns out I'll still have to write a custom strtok because the vanilla one won't parse two consecutive delimiters as a null, which is counterproductive to our use case

Reply:
Can't you use `strsep`?

Reply:
Thanks, changed use this